### PR TITLE
Changes to render thread-safe the package, with test. 

### LINKED ISSUE #58: Thread safety issue.
Working on an unrelated code that uses Perl threads I discovered an issue that seems related to Geo::GDAL::FFI. 

A test script that simply works and includes multiple threads and Thread::Queues, which  does not use any GDAL function, suddenly starts with the following errors by simply adding a 'use Geo::GDAL::Geo' in the code.  I even commented out the DESTROY sub without any luck. It seems not related to FFI::Platypus or other dependencies. I noticed that the BEGIN section of FFI.pm does run a new() . Once commented out, the issue disappears. My hypothesis is that the $instance variable needs a proper protection against threads, which could be done by caller subs eventually, but including new() in the BEGIN sec triggers threads issues.

```
Scalars leaked: -3
Attempt to free unreferenced scalar: SV 0x556a02d45158, Perl interpreter: 0x556a06e4d540 during global destruction.
Attempt to free unreferenced scalar: SV 0x556a02d44f18, Perl interpreter: 0x556a06e4d540 during global destruction.
Attempt to free unreferenced scalar: SV 0x556a02d450b0, Perl interpreter: 0x556a06e4d540 during global destruction.
Aborted
```

### PR DESCRIPTION
**Pull request by fpl**
This patch closes #58 . Even removed the `string_pointer` use which is obsolete since FFI::Platypus 0.61. Added a t/threads.t test to verify that multi-threading does not crash.

Basically, ` CPLPushErrorHandler()`  must be performed per thread, not at `new()` time for the whole program, that is **after** creation of any thread. That implies that there are two possibilities:

- avoid the singleton pattern, so perform the `Geo::GDAL::FFI` initialization for any thread after its creation.
- separate error handling initialization from general initialization.

This patch follows the second approach. Note the use of sticky mode for closure() as required in the documentation. 
Not sure the DESTROY block makes sense, but it seems it does not hurt.

### CODE REVIEW

**Comment by shawnlaffan** [comment]
Works for me on with Strawberry Perl 5.38, and unthreaded perl 5.36 on Ubuntu. 

```t/open.t``` is now noisier although the tests pass.  

```
t/open.t ............... ERROR 4: `test.shp' not recognized as a supported file format.
t/open.t ............... ok
```

**Comment by shawnlaffan** [comment]
It seems the VERBOSE_ERROR output in that test is going straight to stderr.

**Comment by fpl** [comment]
> VERBOSE_ERROR
That's because now
```
 Geo::GDAL::FFI->get_instance->SetErrorHandling;
```
should be used before any other thing. That is the side effect of removing it in the new() sub for thread safety.

**Comment by shawnlaffan** [comment]
Having an explicit ```init``` call might be a reasonable thing to do.

Alternately the ```new``` method could be refactored to not do any of the function bindings, instead only setting the error handler and any other thread specific properties that come up.  I think it would still need a singleton pattern, though, as there should be only one error handler per thread.  

Any of the above changes would be separate to this PR.  

@ajolma - what do you think?

**Comment by ajolma** [comment]
I haven't followed this much so I don't have a very educated opinion. 

One this is that those changes to the attach method calls should be done to build-tools/parse_h.pl, which is used to create them by running against GDAL sources.

**Comment by shawnlaffan** [comment]
parse_h.pl is changed as part of the first commit.

**Comment by fpl** [comment]
> One this is that those changes to the attach method calls should be done to build-tools/parse_h.pl, which is used to create them by running against GDAL sources.

Talking about that, even issue #53, which is managed by PR #68, would require an extensive change to the same build tool, as explained in the same issue.

**Comment by fpl** [comment]
> Having an explicit `init` call might be a reasonable thing to do.
> 
> Alternately the `new` method could be refactored to not do any of the function bindings, instead only setting the error handler and any other thread specific properties that come up. I think it would still need a singleton pattern, though, as there should be only one error handler per thread.
> 
> Any of the above changes would be separate to this PR.
> 
> @ajolma - what do you think?

IMHO, the new() method does the right thing as a singleton. The only setup that needs to be moved out of the BEGIN block is the GDAL error handling that needs to be performed per thread instead. As a general pattern, any lib binding should have a double-step initialization, one for the general setup and one for per thread, to ensure thread safety. Other packages do not have a general setup and require multiple instances (one for each thread) to work. In this case, I think it's a bit too heavy loading all the FFI stuff for every thread by removing the singleton constraint at all, if that is what you suggest.

**Comment by shawnlaffan** [comment]
> IMHO, the new() method does the right thing as a singleton.

The issue is that this would be a breaking change for existing code.  

The FFI attach calls only need to be done once when the module is loaded, so they can be refactored out of the ```new``` sub.  The ```new``` sub could then be called per-thread and set the error handling as it does now.  This means it is essentially the ```init``` call.

Unless of course there are complications due to threads...

**Comment by fpl** [comment]
> > IMHO, the new() method does the right thing as a singleton.
> 
> The issue is that this would be a breaking change for existing code.
> 
> The FFI attach calls only need to be done once when the module is loaded, so they can be refactored out of the `new` sub. The `new` sub could then be called per-thread and set the error handling as it does now. This means it is essentially the `init` call.
> 
> Unless of course there are complications due to threads...

Are you proposing to move the problematic thread unsafe stuff in new() (to be called per thread) and keep all the rest in a new sub init(), to be called in the BEGIN block to initialize the $instance ?

**Comment by fpl** [comment]
@shawnlaffan  It would be nice if someone could review my proposal to split the initialization code into two parts, one for the BEGIN section and one to be called per thread. Of course it would represent a change in the current API, but I can't see any other way to ensure thread safety.

**Comment by shawnlaffan** [comment]
The issue is that the changes break the current API and now non-threaded systems do not have an error handler set by default.  

I wonder if the new call should generate an object that keeps track of which thread it is in.  Then it can set the error handler if the thread field is undef.  That might need a lot more engineering, though, depending on how GDAL does things.

**Comment by fpl** [comment]
> The issue is that the changes break the current API and now non-threaded systems do not have an error handler set by default.
> 
> I wonder if the new call should generate an object that keeps track of which thread it is in. Then it can set the error handler if the thread field is undef. That might need a lot more engineering, though, depending on how GDAL does things.


That's my fault: RTFM applies as often in a dev life.
 
I think a better approach would be de-registering the error handler in a CLONE_SKIP sub and re-registering per thread in a CLONE sub. That would render the module thread-safe and avoid the need to change APIs. Ok, I'm a silly dumb, guys...
I'll change the PR and make some checks, but this approach seems the most reasonable.

**Comment by fpl** [comment]
Ok, this PR uses a CLONE_SKIP to disable GDAL error handling in the main thread, just before creating a new thread, then re-enables it in the created thread after within a CLONE sub. The existing code requires no API change if it is a single-thread application. The only change is that in multi-threaded apps (which never worked until now), the main thread needs to re-enable GDAL error handling before using GDAL and after the creation of all its threads (CLONE_SKIP is called at every thread creation). An alternative would be returning 1 in CALL_SKIP and forcing the creation of a new() instance in every thread before using GDAL.

**Comment by shawnlaffan** [comment]
This passes tests on a local perl 5.38.2 built with ithreads.

**Comment by fpl** [comment]
I would add a note into the FFI pod about re-enabling error handling in the main thread after thread pool creation, if acceptable. That would complete the PR.

**Comment by fpl** [comment]
Add a note in the FFI.pm POD section as said.

**Comment by fpl** [comment]
Hi there, any hope to see this PR considered for next release? This patch is much less intrusive and solves the support of multi-threading app without big problems for existing code (which I guess do not use threads, else it would simply not work at all, as was mine). Even, the drop of the obsolete StringPointer module using is simply a due act and solves even memory leaks under the same MTs conditions. 
Thanks for reviewing, just in case.

**Comment by shawnlaffan** [comment]
@ajolma - I have commit privileges to this repo but I don't think I have PAUSE access to upload to CPAN.

File: lib/Geo/GDAL/FFI.pm [hunk 1]
```diff
@@ -1820,6 +1822,21 @@ BEGIN {
     $instance = Geo::GDAL::FFI->new($gdal);
 }
 
+#
+# The next two subs are required for theread-safety, because GDAL error handling must be set per thread.
```
Comment (shawnlaffan) [inline lib/Geo/GDAL/FFI.pm:1826]: typo: s/theread/thread/

File: lib/Geo/GDAL/FFI.pm [hunk 2]
```diff
@@ -2045,6 +2062,16 @@ confessed if a method fails. This is the default.
 Unset the Perl function to catch GDAL errors. If no other error
 handler is set, GDAL prints the errors into stderr.
 
+=head1 NOTES ABOUT THREAD-SAFETY
+
+This module is mostly thread-safe, but for the error handling management.
```
Comment (shawnlaffan) [inline lib/Geo/GDAL/FFI.pm:2067]: Could this line be deleted once the PR is in?  Or is it best left for future reference? If the latter then some rephrasing would be good.  e.g. "This module is thread safe provided the error handling is taken care of."

File: lib/Geo/GDAL/FFI.pm [hunk 3]
```diff
@@ -2045,6 +2062,16 @@ confessed if a method fails. This is the default.
...
+To ensure thread-safety GDAL error handling is automatically disabled
+before creating a new thread and re-enabled after that in the just
+created thread. The main thread needs to renable it via C<SetErrorHandling>,
+after all thread creations and before eventually using any GDAL function: this
```
Comment (shawnlaffan) [inline lib/Geo/GDAL/FFI.pm:2071]: s/function: this/function. This/

File: lib/Geo/GDAL/FFI.pm [hunk 4]
```diff
@@ -2045,6 +2062,16 @@ confessed if a method fails. This is the default.
...
+must be done explicitly in the main thread, because there is no way
```
Comment (shawnlaffan) [inline lib/Geo/GDAL/FFI.pm:2072]: delete comma in "thread, because"

**Comment by shawnlaffan** [comment]
@fpl - I've added some nit-picky comments to the non-code components above.  I'll have a closer look at the code itself soon.

**Comment by ajolma** [comment]
> @ajolma - I have commit privileges to this repo but I don't think I have PAUSE access to upload to CPAN.

You're a co-maintainer of Geo::GDAL::FFI* modules.

File: lib/Geo/GDAL/FFI.pm (hunk 1 above)
Comment (fpl) [reply lib/Geo/GDAL/FFI.pm]: Done

File: lib/Geo/GDAL/FFI.pm (hunk 2 above)
Comment (fpl) [reply lib/Geo/GDAL/FFI.pm]: Done.

File: lib/Geo/GDAL/FFI.pm (hunk 3 above)
Comment (fpl) [reply lib/Geo/GDAL/FFI.pm]: Done.

File: lib/Geo/GDAL/FFI.pm (hunk 4 above)
Comment (fpl) [reply lib/Geo/GDAL/FFI.pm]: Done

**Comment by shawnlaffan** [comment]
@fpl - the code looks good but unfortunately the recent merger of #70 is now causing merge conflicts. 

Are you able to resolve them?  Ideally the commits in this PR would also be rebased on the current master branch and force-pushed but I can also do that at merge time if needed.

**Comment by shawnlaffan** [comment]
Thanks @fpl - changes merged.  I'll make a few other changes and cut a new release to CPAN.

**Pull request merged**